test/Dialect/FIRRTL/layers-errors.mlir

```diff
@@ -0,0 +1,92 @@
+    // expected-note  @below {{discarding layer requirements: @A}}
+    %s = firrtl.ref.cast %w : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.module @UncoloredDefineUnderLayerblock(out %o : !firrtl.probe<uint<1>>) {
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0  : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.layer @B bind {}
+  firrtl.module @Foo(out %o : !firrtl.probe<uint<1>, @B>) {
+    firrtl.layerblock @B {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @B>
+      firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @B>
+    }
+  }
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>, @B>) {
+    firrtl.layerblock @A {
+      %foo_o = firrtl.instance foo @Foo(out o : !firrtl.probe<uint<1>, @B>)
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %foo_o : !firrtl.probe<uint<1>, @B>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.extmodule @Foo(out o : !firrtl.probe<uint<1>>) attributes {layer = [@A]}
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    %foo_o = firrtl.instance foo @Foo(out o : !firrtl.probe<uint<1>>)
+    firrtl.ref.define %o, %foo_o : !firrtl.probe<uint<1>>
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    %w = firrtl.wire : !firrtl.openbundle<f: probe<uint<1>>>
+    firrtl.layerblock @A {
+      %f = firrtl.opensubfield %w[f] : !firrtl.openbundle<f: probe<uint<1>>>
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %f, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
```

Comment:
This could use at least one `ref.resolve` test.

Reply:
I can add some tests here but they will just be duplicates of preexisting ref.resolve tests in the more general errors.mlir https://github.com/llvm/circt/pull/6643/files#diff-e012a404f9fde3c2af947cbfbf8abeb5c8126f24a6f45d44d530e4923edb0510R1909

I decided not to move any tests around for this PR.